### ISSUE TITLE
[Bug Report] Argmax invalid result with padding

### ISSUE THREAD

**Issue by hschoi4448** [tenstorrent/tt-metal#7571]
**Describe the bug**

When there is padding, the result of argmax is incorrect.

**To Reproduce**
1. Copy below code
```Python
# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.

# SPDX-License-Identifier: Apache-2.0

import torch
import pytest
import tt_lib
from loguru import logger
from tests.tt_eager.python_api_testing.sweep_tests import comparison_funcs


@pytest.mark.parametrize(
    "input_shapes",
    (
        (torch.Size([1, 1, 1, 10])),
    ),
)
@pytest.mark.parametrize(
    "dim",
    (3,),
)
@pytest.mark.parametrize("all", (True,))
class TestArgmax:
    def test_argmax(self, input_shapes, dim, all, device):
        torch.manual_seed(10)

        input_data = torch.tensor([5,3,2,1,10,9,6,7,4,8]).reshape(input_shapes).bfloat16()
        print("input_data", input_data)
        input_tensor = (
            tt_lib.tensor.Tensor(input_data, tt_lib.tensor.DataType.BFLOAT16).pad_to_tile(100).to(tt_lib.tensor.Layout.TILE).to(device)
        )
        tt_output_tensor_on_device = tt_lib.tensor.argmax(input_tensor, dim=dim, all=all)

        tt_out_tensor = tt_output_tensor_on_device.cpu().to(tt_lib.tensor.Layout.ROW_MAJOR).to_torch()
        # tt_out_tensor = tt_output_tensor_on_device.cpu().to(tt_lib.tensor.Layout.ROW_MAJOR)
        if all:
            golden_tensor = torch.argmax(input_data)
            tt_out_tensor = tt_out_tensor[0, 0, 0, 0]
        else:
            golden_tensor = torch.argmax(input_data, dim=dim)
            if dim == 1 or dim == -3 or dim == 0 or dim == -4:
                tt_out_tensor = tt_out_tensor[0]
            else:
                if input_shapes[1] != 1 or input_shapes[0] != 1:
                    tt_out_tensor = tt_out_tensor[0]
                else:
                    tt_out_tensor = tt_out_tensor[0, 0, 0]
            
        pt_out_tensor = golden_tensor
        comp_pass, comp_out = comparison_funcs.comp_pcc(pt_out_tensor, tt_out_tensor, pcc=0.99)
        comp_all, _ = comparison_funcs.comp_allclose(pt_out_tensor, tt_out_tensor, atol=4, rtol=1e-1)
        logger.info(comp_pass)
        logger.info(comp_all)
        logger.info(comp_out)
        status = comp_pass | comp_all
        
        print("pt_out_tensor", pt_out_tensor)
        print("tt_out_tensor", tt_out_tensor)
        assert status

```
2. Check result
```Python
input_data tensor([[[[ 5.,  3.,  2.,  1., 10.,  9.,  6.,  7.,  4.,  8.]]]], dtype=torch.bfloat16)
pt_out_tensor tensor(4)
tt_out_tensor tensor(10., dtype=torch.bfloat16)
```

**Expected behavior**

It seems that when the value of padding is greater than the value of the tensor, argmax returns the index of the padding value.

**Screenshots**

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit a059360c5f356621d37963009ccd5a8089a71432)
- Arch: wormhole_b0

**Additional context**

**Comment by jliangTT** [comment]
Does this one belong to you @umadevimcw , let me know?

**Comment by umadevimcw** [comment]
@jliangTT Yes will check it out

**Comment by umadevimcw** [comment]
@hschoi4448 The problem is with the above logic you have used. Since argmax is to find the maximum index, you should not pad the input with 100. Also, you are sending the Padded input to TT and the unpadded input to Pytorch which is not an apple to apple comparison.  

In the above code either use `-1.0` or `-inf` as pad value. If you change that the test is passing

```
input_tensor = (
            tt_lib.tensor.Tensor(input_data, tt_lib.tensor.DataType.BFLOAT16).pad_to_tile(-1.0).to(tt_lib.tensor.Layout.TILE).to(device)
        )
```

<img width="1507" alt="image" src="https://github.com/tenstorrent/tt-metal/assets/29221091/1db63465-120f-44b9-ab2f-57e2945f7312">

```
`---- ttSiliconDevice::init_hugepage: bind_area_to_memory_nodeset() failed (physical_device_id: 0 ch: 0). Hugepage allocation is not on NumaNode matching TT Device. Side-Effect is decreased Device->Host perf (Issue #893).
                  Metal | INFO     | Initializing device 0. Program cache is NOT enabled
                  Metal | INFO     | AI CLK for device 0 is:   1202 MHz
input_data tensor([[[[ 5.,  3.,  2.,  1., 10.,  9.,  6.,  7.,  4.,  8.]]]],
       dtype=torch.bfloat16)
2024-04-29 07:20:21.422 | INFO     | tests.tt_eager.python_api_testing.sweep_tests.pytests.tt_dnn.test_argmax:test_argmax:52 - True
2024-04-29 07:20:21.422 | INFO     | tests.tt_eager.python_api_testing.sweep_tests.pytests.tt_dnn.test_argmax:test_argmax:53 - True
2024-04-29 07:20:21.422 | INFO     | tests.tt_eager.python_api_testing.sweep_tests.pytests.tt_dnn.test_argmax:test_argmax:54 - Max ATOL Delta: 0.0, Max RTOL Delta: 0.0, PCC: 1.0
pt_out_tensor tensor(4)
tt_out_tensor tensor(4., dtype=torch.bfloat16)
PASSED                  Metal | INFO     | Closing device 0
                  Metal | INFO     | Disabling and clearing program cache on device 0


================================================================= PASSES =================================================================
======================================================== short test summary info =========================================================
PASSED tests/tt_eager/python_api_testing/sweep_tests/pytests/tt_dnn/test_argmax.py::TestArgmax::test_argmax[True-3-input_shapes0]
=========================================================== 1 passed in 8.13s ============================================================
                 Device | INFO     | Closing user mode device drivers`
```

**Comment by davorchap** [comment]
@umadevimcw 
the request for argmax is to be able to work with padded tensors and to ignore the padding, that is argmax only on the unpadded regionof the tensor, ie the orignal shape of the tensor should be used to determine which values of the padded tensor are considered in argmax.
This can be done by argmax internally masking the padded region with -inf, within the CB. This will allow it to work with any padding, even if it's 100 or garbage.

**Comment by ruthreshx** [comment]
@davorchap May I know which file I have to take a look in GroupNorm?

**Comment by umadevimcw** [comment]
@davorchap By default we have `-inf` as padding for the reduce max op. On top of it do you want to mask the padded value and work? Since the user will pad the files, do you have any specific method to differentiate between padded and unpadded shape while processing the Tensor?

**Comment by umadevimcw** [comment]
@davorchap @hschoi4448 Do you want us to test only for the [1,1,1,10] this kind of tensors? Is it always N, C, H values will always be 1 ?

**Comment by hschoi4448** [comment]
>@davorchap @hschoi4448 Do you want us to test only for the [1,1,1,10] this kind of tensors? Is it always N, C, H values will always be 1 ?

as far as I know, in MNIST, the argmax input tensor size is (BatchSize, 10). Since we cannot accept a 2D tensor as input, I am planning to use a tensor of size (1, 1, BatchSize, 10).

**Comment by umadevimcw** [comment]
@hschoi4448 #8202 This PR holds the masking of user pad value implementation. I have added test files also.
when the shapes are 1,1,10,10 -> after padding 1,1,32,32, the index return by the implementation will account the padded shape. In such cases, we must send a padded input to the torch. 

Please test and let me know, so that I can merge it to main branch : `umadevimcw/argmax_with_padding`

**Comment by hschoi4448** [comment]
Recently, during testing, I discovered an issue where the output dtype of argmax is not an integer type. When inspecting `torch.argmax`, the output dtype is `LongTensor`, but as TT currently lacks a long type, it seems like it should return a `signed integer type` for now.

Would you prefer to address this issue within this current one, or would it be better for me to create a separate issue for this?"

@davorchap @umadevimcw @razorback3

**Comment by umadevimcw** [comment]
@tt-aho https://github.com/tenstorrent/tt-metal/issues/7571#issuecomment-2114068845 any comments on this?

@hschoi4448 Do you want the argmax output in Integer format?

**Comment by razorback3** [comment]
> @hschoi4448 Do you want the argmax output in Integer format?

Yes, as following the torch spec 😄

**Comment by umadevimcw** [comment]
@hschoi4448 Can you retest this PR #8202 (it doesn't include the int format output)?

**Comment by hschoi4448** [comment]
> @hschoi4448 Can you retest this PR #8202 (it doesn't include the int format output)?

I have run the tests on the WH equipment and confirmed that they passed. 👍

**Comment by umadevimcw** [comment]
@hschoi4448 I have merged the PR #8202 hence closing this issue. Kindly create an separate issue for the int format support
